feat: GitHub Actions improvements (auto-release, better tests)

smallstepman opened the pull request:
- Additional check if `assets.did` file has the same content as the upstream one https://github.com/dfinity/cdk-rs/blob/main/src/ic-certified-assets/assets.did
- Automatic releases when pushing to main ([example](https://github.com/smallstepman/certified-assets/releases/tag/0.2.2))

smallstepman commented:
Check and Test is supposed to fail until this #31 is merged

ericswanson-dfinity commented on .github/workflows/cicd.yml:
```diff
@@ -0,0 +1,76 @@
+      - name: Compute diff between local and upstream candid files
+        run: |
+          curl -s https://raw.githubusercontent.com/dfinity/cdk-rs/main/src/ic-certified-assets/assets.did > upstream_assets.did
+          echo CANDID_DIFF=$(diff upstream_assets.did assets.did | wc -l) >> $GITHUB_ENV
```
What would you think about displaying the diff here, and capturing the return code (which will be nonzero if there were differences) for later comparison?  Something like
```suggestion
          diff upstream_assets.did assets.did
          echo CANDID_DIFF_RC="$?" >> $GITHUB_ENV
```

This will help troubleshooting when there is a difference.

smallstepman replied:
btw, what does `_RC` stand for here?

ericswanson-dfinity replied:
Return Code.  But name it what you want: could be `CANDID_DIFF_RESULT` or something else!

ericswanson-dfinity commented on .github/workflows/ci.yml:
```diff
@@ -38,4 +38,17 @@ jobs:
+        with:
+          script: |
+              core.setFailed('Candid file for asset canister is different from upstream (dfinity/cdk-rs/ic-certified-assets)')
+              
```
I recommend to make the workflow report these in a status that does not depend on the matrix.  This allows us to set up branch protection rules that don't have to change when the rust version changes.

Example: https://github.com/dfinity/cdk-rs/blob/main/.github/workflows/e2e.yml#L55

Alternately, I'd suggest to make a separate workflow for this check, since it doesn't depend on rust and won't need a matrix at all.

smallstepman replied:
I wanted to avoid installing cargo twice, but you're right, that's the correct approach. 

I have this good-bad habit where I'll try to save computational resources to reduce carbon emission - a drop in the ocean, which makes no sense here.

smallstepman replied:
disregard what I said about cargo - my mind was in another GH Action yml at the time of writing...

smallstepman replied:
I moved this step into separate workflow `ci-candid.yml`

smallstepman commented:
I took another stab at it:

- split candid check into separate yml
- use actions-rs/toolchain instead of manually installing rust toolchain (uses rust-toolchain.toml) (had to use forked version, let me know if how big of a no-no it is from security standpoint. I already emailed the author of actions-rs/toolchain asking to add more maintainers to the project) 
- make release workflow run only on merged pull request to main banch
- split dfx install and wasm tasks in release workflow
- shave few tasks from release workflow
- use rust toolchain version from rust-toolchain.toml for workflow caching actions/cache

ericswanson-dfinity commented on .github/workflows/ci-candid.yml:
```diff
@@ -0,0 +1,19 @@
+name: Compare candid spec with upstream
+on: [push]
```
should we do this also on pull_request?

smallstepman replied:
totally

ericswanson-dfinity reviewed (commented):
This looks good!  The only thing I would recommend is adding name fields to the jobs, following a convention, in order to easily identify them for inclusion in branch protection rules.

For example, see https://github.com/dfinity/agent-rs/pull/348, which has these required checks:
- fmt:required
- ic-ref:required
- icx-asset:required
- lint:required

I'd suggest calling these two `candid:required` and `rust:required`

smallstepman commented:
Thanks!

I've never set up branch protection rules myself, so I'm don't have a well-established feeling for how it works or what's the good naming convention or which candidates are a good pick for `:required`, so for now I'm just going to follow your suggestion to see how this plays out, but overall the idea looks sweet, tl;dr many thanks for suggestion:D.

Done
